File: src/compiler/parser/PredictiveParserException.java
package compiler.parser;

import compiler.lexer.token.Token;

import java.util.stream.Collectors;

public class PredictiveParserException extends UserException {

  public PredictiveParserException(GrammarNode grammarNode, Token token) {
    super(formatMessage(grammarNode, token), token);
  }

  private static String formatMessage(GrammarNode grammarRule, Token token) {
    String tokenClassName = token.getClass().getSimpleName();
    String grammarClassName = grammarRule.getClass().getSimpleName();
    String expectedTokens = grammarRule.getRHS().stream().map(Class::getSimpleName).collect(Collectors.joining(" or "));
    return "\n" +
      "LL Table missing entry exception; " + grammarRule + "(" + tokenClassName + ") = undefined" +
      "\n" +
      grammarClassName + " expected " + expectedTokens + " but found " + tokenClassName;
  }
}

File: src/compiler/parser/PredictiveParserException.kt
package compiler.parser

import compiler.parser.Language.Grammar
import compiler.parser.Language.Token
import java.util.stream.Collectors

class PredictiveParserException(grammarNode: Grammar, token: Token) : UserException(formatMessage(grammarNode, token), token) {
    companion object {
        private fun formatMessage(grammarRule: Grammar, token: Token): String {
            val tokenClassName = token.javaClass.simpleName
            val grammarClassName = grammarRule.javaClass.simpleName
            val expectedTokens = grammarRule
                    .getRHS()
                    .stream()
                    .map { it.simpleName }
                    .collect(Collectors.joining(" or "))
            return "\nLL Table missing entry exception; $grammarRule($tokenClassName) = undefined\n$grammarClassName expected $expectedTokens but found $tokenClassName"
        }
    }
}
